hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/CassAlertsServiceImpl.java

```diff
@@ -0,0 +1,849 @@
+
+        wakeUpTimer = new Timer("CassAlertsServiceImpl-Timer");
+
+        delay = new Integer(AlertProperties.getProperty(ENGINE_DELAY, "1000"));
+        period = new Integer(AlertProperties.getProperty(ENGINE_PERIOD, "2000"));
+    }
+
+    @PostConstruct
+    public void initServices() {
+        try {
+            if (this.keyspace == null) {
+                this.keyspace = AlertProperties.getProperty(CASSANDRA_KEYSPACE, "hawkular_alerts");
+            }
+
+            if (session == null) {
+                session = CassCluster.getSession();
+            }
+
+            initPreparedStatements();
+
+            GsonBuilder gsonBuilder = new GsonBuilder();
+            gsonBuilder.registerTypeHierarchyAdapter(ConditionEval.class, new GsonAdapter<ConditionEval>());
+            gson = gsonBuilder.create();
+
+            reload();
+        } catch (Throwable t) {
+            msgLog.errorCannotInitializeAlertsService(t.getMessage());
+        }
+    }
+
+    private void initPreparedStatements() throws Exception {
+        if (insertAlert == null) {
+            insertAlert = session.prepare("INSERT INTO " + keyspace + ".alerts " +
+                    "(alertId, payload) VALUES (?, ?) ");
+        }
+        if (insertAlertTrigger == null) {
+            insertAlertTrigger = session.prepare("INSERT INTO " + keyspace + ".alerts_triggers " +
+                    "(alertId, triggerId) VALUES (?, ?) ");
+        }
+        if (insertAlertCtime == null) {
+            insertAlertCtime = session.prepare("INSERT INTO " + keyspace + ".alerts_ctimes " +
+                    "(alertId, ctime) VALUES (?, ?) ");
+
+        }
+        if (insertAlertStatus == null) {
+            insertAlertStatus = session.prepare("INSERT INTO " + keyspace + ".alerts_statuses " +
+                    "(alertId, status) VALUES (?, ?) ");
+        }
+        if (updateAlert == null) {
+            updateAlert = session.prepare("UPDATE " + keyspace + ".alerts " +
+                    "SET payload = ? WHERE alertId = ? ");
+        }
+        if (selectAlertStatus == null) {
+            selectAlertStatus = session.prepare("SELECT alertId, status FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? ALLOW FILTERING ");
+        }
+        if (deleteAlertStatus == null) {
+            deleteAlertStatus = session.prepare("DELETE FROM " + keyspace + ".alerts_statuses " +
+                    "WHERE alertId = ? AND status = ?");
+        }
+    }
+
+    public void addAlerts(Collection<Alert> alerts) throws Exception {
+        if (alerts == null) {
+            throw new IllegalArgumentException("Alerts must be not null");
+        }
+        if (session == null) {
+            throw new RuntimeException("Cassandra session is null");
+        }
+        if (insertAlert == null) {
+            throw new RuntimeException("insertAlert PreparedStatement is null");
+        }
+        try {
+            for (Alert a : alerts) {
```

Comment:
Since the driver is async, it is typically better to do these inserts in parallel something like,

```java
List<ResultSetFuture> futures = new ArrayList<>;
for (Alert a : alerts) {
    futures.add(session.executeAsync(insertAlert.bind(a.getAlertId(), toJson(a)));
    futures.add(session.executeAsync(insertAlertTrigger.bind(a.getAlertId(), a.getTriggerId())));
    futures.add(session.executeAsync(insertAlertCtime.bind(a.getAlertId(), a.getCtime())));
    futures.add(session.executeAsync(insertAlertStatus.bind(a.getAlertId(), a.getStatus().name())));
}
// If any of the futures fail, the future returned from Futures.allAsList will fail. This is a somewhat simplistic approach to error handling, particularly if you are inserting a lot of alerts, but it is a good starting point. If the returned future fails, you simply retry the inserts. 
return Futures.allAsList(futures);
```

Doing async, parallel inserts will be much faster and more scalable  than doing blocking inserts particularly as the number of inserts increases. I might also be inclined to further refactor my code to eliminate some of the boiler plate, creating some reusable functions in the process,

```java
Function<List<BoundStatement>, List<ResultSetFuture>> insertAsync = statements ->
                statements.stream().map(session::executeAsync).collect(toList());

Function<Alert, List<BoundStatement>> bindAlert = alert -> asList(
                insertAlert.bind(alert.getAlertId(), toJson(alert)),
                insertAlertTrigger.bind(alert.getAlertId(), alert.getTriggerId()),
                insertAlertCtime.bind(alert.getAlertId(), alert.getCtime()),
                insertAlertStatus.bind(alert.getAlertId(), alert.getStatus().name()));

return Futures.allAsList(alerts.stream()
                .map(bindAlert)
                .map(insertAsync)
                .flatMap(List::stream)
                .collect(toList()));
```